sklearn/metrics/pairwise.py

```diff
@@ -935,13 +939,47 @@ def _parallel_pairwise(X, Y, func, n_jobs, **kwds):
     if Y is None:
         Y = X
 
+    if n_jobs == 1:
+        return func(X, Y, **kwds)
+
+    # TODO: in some cases, backend='threading' may be appropriate
+    fd = delayed(func)
     ret = Parallel(n_jobs=n_jobs, verbose=0)(
-        delayed(func)(X, Y[s], **kwds)
+        fd(X, Y[s], **kwds)
         for s in gen_even_slices(Y.shape[0], n_jobs))
 
     return np.hstack(ret)
 
 
+def _pairwise_callable(X, Y, metric, **kwds):
+    """Handle the callable case for pairwise_{distances,kernels}
+    """
+    X, Y = check_pairwise_arrays(X, Y)
+    if X is Y:
+        # Only calculate metric for upper triangle
+        out = np.zeros((X.shape[0], Y.shape[0]), dtype='float')
+        iterator = itertools.combinations(range(X.shape[0]), 2)
+        iterator = itertools.combinations(range(X.shape[0]), 2)
+    else:
+        # Calculate all cells
+        out = np.empty((X.shape[0], Y.shape[0]), dtype='float')
+        iterator = itertools.product(range(X.shape[0]), range(Y.shape[0]))
+
+    for i, j in iterator:
+        out[i, j] = metric(X[i], Y[j], **kwds)
+
+    if X is Y:
+        # Make symmetric
+        # NB: out += out.T will produce incorrect results
```

Comment:
Indeed, it is very important to keep in mind that the idiome that you are trying to use is incorrect, because out.T is a view on 'out'.

In practice, for reasons of buffering that I do not understand, it produces incorrect results only if the out matrix is large enough. This is very nasty, because it means that people do not see the problem when simply playing around to test the code.